Unbacked SymInt support for pointwise and multiple boolean masks

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95222
* #95218
* __->__ #94790
* #94523
* #95209
* #95208

This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted` and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Unbacked SymInt support for CrystalDPR

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for CrystalDPR"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for CrystalDPR"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ngimel commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -583,13 +583,23 @@ def infer_size(a, b):
         sizeB = b[dimB] if dimB >= 0 else 1
-        if not (sizeA == sizeB or sizeA == 1 or sizeB == 1):
+        from torch.fx.experimental.symbolic_shapes import definitely_true
+        if definitely_true(sizeA == 1):
```
can you leave a comment here about how `definitely_true` case is different from `sizeB == 1` below? We don't expect unbacked symints to hit `sizeB == 1`?

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Unbacked SymInt support for pointwise and multiple boolean masks"


This PR adds support for some more unbacked SymInt cases extracted from CrystalDPR

* You can do pointwise operations with tensors that have unbacked SymInts; fixes involve making broadcasting avoid touching unbacked SymInts if not necessary (in both PrimTorch and the fake tensor fastpath), switching PrimTorch decomp for elementwise operations to use `empty_permuted and short circuiting if inputs are contiguous. I added `definitely_true` and `parallel_or` helpers to handle this case elegantly; see docblocks for more discussion.
  * The `is_contiguous` change requires me to update PrimTorch to let `is_contiguous` through directly, which seems fine.
* Add a new `constrain_unify` helper which lets you specify that two unbacked SymInts are the same. This addresses the case when you use the same boolean mask to index into different tensors; the resulting slices have the same shapes but we do not currently know this statically.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
